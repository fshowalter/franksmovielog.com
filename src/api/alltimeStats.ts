import type { AlltimeStatsJson } from "./data/alltimeStatsJson";

import { alltimeStatsJson } from "./data/alltimeStatsJson";

export type AlltimeStats = {} & AlltimeStatsJson;

export async function alltimeStats(): Promise<AlltimeStats> {
  return await alltimeStatsJson();
}
