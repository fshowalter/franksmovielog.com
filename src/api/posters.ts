import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export type PosterImageProps = {
  src: string;
  srcSet: string;
};

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/posters/*.png",
);

export async function getFixedWidthPosterImageProps(
  slug: null | string,
  {
    height,
    width,
  }: {
    height: number;
    width: number;
  },
): Promise<PosterImageProps> {
  if (!slug) {
    slug = "default";
  }

  const posterFilePath = Object.keys(images).find((path) => {
    return path.endsWith(`/${slug}.png`);
  })!;

  const posterFile = await images[posterFilePath]();

  const optimizedImage = await getImage({
    densities: [1, 2],
    format: "avif",
    height: height,
    quality: 80,
    src: posterFile.default,
    width: width,
  });

  return {
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
  };
}

export async function getFluidWidthPosterImageProps(
  slug: string | undefined = "default",
  {
    height,
    width,
  }: {
    height: number;
    width: number;
  },
): Promise<PosterImageProps> {
  const posterFilePath = Object.keys(images).find((path) => {
    return path.endsWith(`/${slug}.png`);
  })!;

  const posterFile = await images[posterFilePath]();

  const optimizedImage = await getImage({
    format: "avif",
    height: height,
    quality: 80,
    src: posterFile.default,
    width: width,
    widths: [0.25, 0.5, 1, 2].map((w) => w * width),
  });

  return {
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
  };
}
