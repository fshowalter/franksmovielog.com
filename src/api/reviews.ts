import rehypeRaw from "rehype-raw";
import rehypeStringify from "rehype-stringify";
import { remark } from "remark";
import remarkGfm from "remark-gfm";
import remarkRehype from "remark-rehype";
import smartypants from "remark-smartypants";
import strip from "strip-markdown";

import type { ReviewedTitleJson } from "./data/reviewedTitlesJson";
import type { MarkdownReview } from "./data/reviewsMarkdown";
import type { MarkdownViewing } from "./data/viewingsMarkdown";

import { allReviewedTitlesJson } from "./data/reviewedTitlesJson";
import { allReviewsMarkdown } from "./data/reviewsMarkdown";
import { allViewingsMarkdown } from "./data/viewingsMarkdown";
import { linkReviewedTitles } from "./utils/linkReviewedTitles";
import { getHtml } from "./utils/markdown/getHtml";
import { removeFootnotes } from "./utils/markdown/removeFootnotes";
import { rootAsSpan } from "./utils/markdown/rootAsSpan";
import { trimToExcerpt } from "./utils/markdown/trimToExcerpt";

let cachedViewingsMarkdown: MarkdownViewing[];
let cachedMarkdownReviews: MarkdownReview[];
let cachedReviewedTitlesJson: ReviewedTitleJson[];
let cachedReviews: Reviews;

if (import.meta.env.MODE !== "development") {
  cachedViewingsMarkdown = allViewingsMarkdown();
  cachedReviewedTitlesJson = await allReviewedTitlesJson();
  cachedMarkdownReviews = allReviewsMarkdown();
}

export type Review = MarkdownReview & ReviewedTitleJson & {};

export type ReviewContent = {
  content: string | undefined;
  excerptPlainText: string;
  viewings: ReviewViewing[];
};

export type ReviewExcerpt = {
  excerpt: string;
};

type Reviews = {
  distinctGenres: string[];
  distinctReleaseYears: string[];
  distinctReviewYears: string[];
  reviews: Review[];
};

type ReviewViewing = MarkdownViewing & {
  mediumNotes: string | undefined;
  venueNotes: string | undefined;
  viewingNotes: string | undefined;
};

export async function allReviews(): Promise<Reviews> {
  if (cachedReviews) {
    return cachedReviews;
  }
  const reviewedTitlesJson =
    cachedReviewedTitlesJson || (await allReviewedTitlesJson());
  const reviews = parseReviewedTitlesJson(reviewedTitlesJson);

  if (!import.meta.env.DEV) {
    cachedReviews = reviews;
  }

  return reviews;
}

export function getContentPlainText(rawContent: string): string {
  return getMastProcessor()
    .use(removeFootnotes)
    .use(strip)
    .processSync(rawContent)
    .toString();
}

export async function loadContent<
  T extends { imdbId: string; rawContent: string; title: string },
>(review: T): Promise<ReviewContent & T> {
  const viewingsMarkdown = cachedViewingsMarkdown || allViewingsMarkdown();
  const reviewedTitlesJson = await allReviewedTitlesJson();

  const excerptPlainText = getMastProcessor()
    .use(removeFootnotes)
    .use(trimToExcerpt)
    .use(strip)
    .processSync(review.rawContent)
    .toString();

  const viewings = viewingsMarkdown
    .filter((viewing) => {
      return viewing.imdbId === review.imdbId;
    })
    .reverse()
    .map((viewing) => {
      return {
        ...viewing,
        mediumNotes: getHtmlAsSpan(viewing.mediumNotesRaw, reviewedTitlesJson),
        venueNotes: getHtmlAsSpan(viewing.venueNotesRaw, reviewedTitlesJson),
        viewingNotes: getHtml(viewing.viewingNotesRaw, reviewedTitlesJson),
      };
    });

  return {
    ...review,
    content: getHtml(review.rawContent, reviewedTitlesJson),
    excerptPlainText,
    viewings,
  };
}

export function loadExcerptHtml<T extends { slug: string }>(
  review: T,
): ReviewExcerpt & T {
  const reviewsMarkdown = cachedMarkdownReviews || allReviewsMarkdown();

  const { rawContent } = reviewsMarkdown.find((markdown) => {
    return markdown.slug === review.slug;
  })!;

  let excerptHtml = getMastProcessor()
    .use(removeFootnotes)
    .use(trimToExcerpt)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeRaw)
    .use(rehypeStringify)
    .processSync(rawContent)
    .toString();

  excerptHtml = excerptHtml.replace(/\n+$/, "");
  excerptHtml = excerptHtml.replace(
    /<\/p>$/,
    ` <a class="!no-underline uppercase whitespace-nowrap font-normal font-sans text-accent text-xs  leading-none hover:!underline" href="/reviews/${review.slug}/">Continue reading...</a></p>`,
  );

  return {
    ...review,
    excerpt: excerptHtml,
  };
}

export async function mostRecentReviews(limit: number) {
  const reviewedTitlesJson =
    cachedReviewedTitlesJson || (await allReviewedTitlesJson());

  reviewedTitlesJson.sort((a, b) => b.sequence.localeCompare(a.sequence));
  const slicedTitles = reviewedTitlesJson.slice(0, limit);

  const { reviews } = parseReviewedTitlesJson(slicedTitles);

  return reviews;
}

function getHtmlAsSpan(
  content: string | undefined,
  reviewedTitles: { imdbId: string; slug: string }[],
) {
  if (!content) {
    return;
  }

  const html = getMastProcessor()
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeRaw)
    .use(rootAsSpan)
    .use(rehypeStringify)
    .processSync(content)
    .toString();

  return linkReviewedTitles(html, reviewedTitles);
}

function getMastProcessor() {
  return remark().use(remarkGfm).use(smartypants);
}

function parseReviewedTitlesJson(
  reviewedTitlesJson: ReviewedTitleJson[],
): Reviews {
  const distinctReviewYears = new Set<string>();
  const distinctReleaseYears = new Set<string>();
  const distinctGenres = new Set<string>();
  const reviewsMarkdown = cachedMarkdownReviews || allReviewsMarkdown();

  const reviews = reviewedTitlesJson.map((title) => {
    for (const genre of title.genres) distinctGenres.add(genre);
    distinctReleaseYears.add(title.year);

    const { date, grade, rawContent } = reviewsMarkdown.find(
      (reviewsmarkdown) => {
        return reviewsmarkdown.slug === title.slug;
      },
    )!;

    distinctReviewYears.add(
      date.toLocaleDateString("en-US", {
        timeZone: "UTC",
        year: "numeric",
      }),
    );

    return {
      ...title,
      date,
      grade,
      rawContent,
    };
  });

  return {
    distinctGenres: [...distinctGenres].toSorted(),
    distinctReleaseYears: [...distinctReleaseYears].toSorted(),
    distinctReviewYears: [...distinctReviewYears].toSorted(),
    reviews,
  };
}
