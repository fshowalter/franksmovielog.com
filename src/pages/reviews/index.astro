---
import { getProps } from "~/components/Reviews/getProps";
import { Reviews } from "~/components/Reviews/Reviews";
import Layout from "~/layouts/Layout.astro";

const props = await getProps();
---

<Layout
  article={false}
  canonical={true}
  description={props.deck}
  title="Reviews"
>
  <Reviews client:load {...props} />
</Layout>
