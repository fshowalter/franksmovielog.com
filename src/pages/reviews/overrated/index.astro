---
import React from "react";

import { Backdrop, BreadcrumbLink } from "~/components/Backdrop";
import { Layout as ComponentLayout } from "~/components/Layout";
import { getPropsForOverrated as getProps } from "~/components/Reviews/getProps";
import { Overrated } from "~/components/Reviews/Overrated";
import { ReviewsSubNav } from "~/components/Reviews/ReviewsSubNav";
import Layout from "~/layouts/Layout.astro";

const props = await getProps();

// Separate layout props from interactive props
const { backdropImageProps, deck, ...interactiveProps } = props;

const breadcrumb = React.createElement(BreadcrumbLink, {
  children: "Reviews",
  href: "/reviews/",
});
---

<Layout
  canonical={true}
  meta={{
    description: props.metaDescription,
  }}
  openGraph={{
    description: deck,
    type: "website",
  }}
  title="Overrated Disappointments"
>
  <ComponentLayout addGradient className="bg-subtle">
    <Backdrop
      breadcrumb={breadcrumb}
      deck={deck}
      imageProps={backdropImageProps}
      title="Overrated"
    />
    <ReviewsSubNav active="overrated" />
    <Overrated client:load {...interactiveProps} />
  </ComponentLayout>
</Layout>
