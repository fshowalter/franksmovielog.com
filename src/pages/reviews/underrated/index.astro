---
import { getProps } from "~/components/Underrated/getProps";
import { Underrated } from "~/components/Underrated/Underrated";
import Layout from "~/layouts/Layout.astro";

const props = await getProps();
---

<Layout
  canonical={true}
  meta={{ description: props.metaDescription }}
  openGraph={{ description: props.deck, type: "website" }}
  title="Underrated Surprises"
>
  <Underrated client:load {...props} />
</Layout>
