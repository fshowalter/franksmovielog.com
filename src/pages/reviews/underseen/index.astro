---
import { getProps } from "src/components/Underseen/getProps";
import { Underseen } from "src/components/Underseen/Underseen";
import Layout from "src/layouts/Layout.astro";

const props = await getProps();
---

<Layout title="Underseen Gems" description={props.deck}>
  <Underseen client:load {...props} />
</Layout>
