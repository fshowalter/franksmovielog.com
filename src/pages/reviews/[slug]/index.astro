---
import { allReviews, getContentPlainText } from "~/api/reviews";
import { getProps } from "~/features/review/getProps";
import { Review } from "~/features/review/Review";
import Layout from "~/layouts/Layout.astro";

interface Props {
  contentPlainText: string;
  slug: string;
}

export async function getStaticPaths() {
  const { reviews } = await allReviews();

  return await Promise.all(
    reviews.map(async (review) => {
      return {
        params: {
          slug: review.slug,
        },
        props: {
          contentPlainText: getContentPlainText(review.rawContent),
          slug: review.slug,
        },
      };
    }),
  );
}

const { contentPlainText, slug } = Astro.props;

const props = await getProps(slug);

//trim the string to the maximum length
var description = contentPlainText
  .replaceAll(/\r?\n|\r/g, " ")
  .slice(0, Math.max(0, 160));

//re-trim if we are in the middle of a word
description = description.slice(
  0,
  Math.max(0, Math.min(description.length, description.lastIndexOf(" "))),
);

let title = `${props.value.title} movie review`;

if (props.value.title.length <= 10) {
  title = `${title} & summary`;
}

title = `${title} (${props.value.releaseYear})`;
---

<Layout
  canonical={true}
  meta={{ description }}
  openGraph={{ description, type: "article" }}
  title={title}
>
  <Fragment slot="meta-theme-color">
    <meta
      content="#fff"
      media="(prefers-color-scheme: light)"
      name="theme-color"
    />
    <meta
      content="#322f2f"
      media="(prefers-color-scheme: dark)"
      name="theme-color"
    />
  </Fragment>
  <Review {...props} />
</Layout>
