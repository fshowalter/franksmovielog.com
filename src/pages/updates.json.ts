import { getUpdatePosterProps } from "~/api/posters";
import { mostRecentReviews } from "~/api/reviews";

const gradeToStars: Record<string, number> = {
  A: 5,
  "A+": 5,
  "A-": 4.5,
  B: 4,
  "B+": 4,
  "B-": 3.5,
  C: 3,
  "C+": 3,
  "C-": 2.5,
  D: 2,
  "D+": 2,
  "D-": 1.5,
  F: 1,
  "F+": 1,
  "F-": 0.5,
};

export async function GET() {
  const reviews = await mostRecentReviews(6);

  const updateItems = await Promise.all(
    reviews.map(async (review) => {
      const posterProps = await getUpdatePosterProps(review.slug);

      return {
        date: review.reviewDate,
        image: posterProps.src,
        slug: review.slug,
        stars: gradeToStars[review.grade],
        title: review.title,
        year: review.releaseYear,
      };
    }),
  );

  return new Response(JSON.stringify(updateItems));
}
