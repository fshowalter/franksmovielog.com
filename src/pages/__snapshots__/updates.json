[
  {
    "date": "2024-04-27T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Fthe-midnight-meat-train-2008.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "the-midnight-meat-train-2008",
    "stars": 2,
    "title": "The Midnight Meat Train",
    "year": "2008"
  },
  {
    "date": "2024-02-18T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Fthe-entity-1982.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "the-entity-1982",
    "stars": 2,
    "title": "The Entity",
    "year": "1982"
  },
  {
    "date": "2023-08-31T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Faqua-teen-forever-plantasm-2022.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "aqua-teen-forever-plantasm-2022",
    "stars": 3,
    "title": "Aqua Teen Forever: Plantasm",
    "year": "2022"
  },
  {
    "date": "2023-07-31T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Fbig-jim-mclain-1952.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "big-jim-mclain-1952",
    "stars": 3,
    "title": "Big Jim McLain",
    "year": "1952"
  },
  {
    "date": "2023-06-23T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Fthe-heavenly-body-1944.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "the-heavenly-body-1944",
    "stars": 1.5,
    "title": "The Heavenly Body",
    "year": "1944"
  },
  {
    "date": "2022-10-06T00:00:00.000Z",
    "image": "/_image/?href=%2Fcontent%2Fassets%2Fposters%2Fthe-oklahoma-kid-1939.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&w=500&h=750&q=100&f=png",
    "slug": "the-oklahoma-kid-1939",
    "stars": 2.5,
    "title": "The Oklahoma Kid",
    "year": "1939"
  }
]
